### ISSUE TITLE
WP: new @qualety/ml plugin — package scaffold + determinism/seeding rules

### ISSUE THREAD

**Issue by NavehBrenner** [NavehBrenner/qualety#117]
Locked SoT — Naveh approved room defaults 2026-09-02. Body is source of truth for planner/build. Do not put planner trigger tokens in this body.

Base main tip at lock: `31244658142ba45f8c72005acded331b790fd552` (v0.1.4).

## Goal

Ship **`@qualety/ml`** as a new official product plugin and implement its first rule bag: **determinism / seeding**, derived from real incidents in NavehBrenner/kevin (unseeded training for months; seed only reached split; worker RNG; silent TF32; missing weight-identity test).

One PR completes this issue (`Closes #117`).

## Non-goals / out of scope

- #118 run provenance, #119 content-hash/cache, #120 train/serve skew, #121 framework misuse — stay stubs; no rules from those here
- #123 threshold-constant / single-source threshold — not ml; leave alone
- #108 abbreviations, #95 deferred TS, #53 multi-lang host
- Root monorepo dogfood enablement of `@qualety/ml` (no torch training tree here)
- `qualety init` auto-adding `@qualety/ml` (install ≠ enable; no torch detect this WP)
- New artifact provider; no second Python frontend; no Ruff/Biome deltas in ml
- Version bump / publish / `/release` (not this PR)
- SPECS multi-lang plugin authoring; plugins remain TS-authored (SPECS locked #4)

## Package / engine shape (locked)

- New workspace package **`packages/ml`** → npm name **`@qualety/ml`**, public publishable like `@qualety/python` / `@qualety/dry` (not private like `@qualety/dev`).
- **`Plugin.name`: `"ml"`** so rule ids are `ml/…`.
- **TS-authored** plugin. Rules use `defineRule` with **`requires: ["python"]`** and `context.getArtifact("python")`.
- **Do not** fork or re-provide the python artifact. Consumer must list **`@qualety/python`** (or another provider of `"python"`) in `plugins[]` when enabling ml rules that need it. Missing provider → existing fail-closed exit 2.
- Package depends on `qualety` workspace; may depend on `@qualety/python` only if needed for types/helpers — prefer not importing python package internals; walk AST via the artifact the same way other external plugins would (local small walk helpers OK, copied/adapted patterns from python plugin are fine if kept inside ml; do not create a hard runtime coupling that breaks when only ml is resolved incorrectly).
- Wire **official loader** in `packages/qualety/src/engine.ts` `officialLoaders` for `"@qualety/ml"`.
- Root workspace `package.json` devDependency / workspace dep on `@qualety/ml` as other official plugins.
- **Companion pair:** `docs/rulesets/ml.md` ↔ `packages/ml/src/index.ts` in `companion-closure.ts`.
- Honest catalog doc **`docs/rulesets/ml.md`**: plugin name, provides nothing, requires python via rules, recommended table, per-rule sections matching implementation (docs↔export honesty).
- SPECS / VISION / api.md / README only as needed for “official plugin exists” honesty — no fake rule claims beyond what ships. Prefer minimal SPECS touch: note `@qualety/ml` under plugin ownership / catalog if there is an existing list; do not rewrite architecture.
- Fixtures under `packages/ml/fixtures/…` + unit tests per rule (valid + invalid, assert rule ids/messages). Concrete suggestions (not `NO_SUGGESTION`). Underapproximate — silence when uncertain.
- Skills `create-plugin` / docs patterns: follow existing python/dry package layout (`package.json` exports dist, `tsconfig`, LICENSE, `src/index.ts` default export plugin).

### Config story for consumers

```json
{
  "plugins": ["@qualety/python", "@qualety/ml"]
}
```

Loading ml applies `configs.recommended` below. User `config.rules` overlays. Installing the package without `plugins[]` enables nothing.

## Training-module signal (shared, locked)

A Python source counts as a **training module** when the file (non-skipped path) statically shows **any** of:

1. A call to an attribute/name **`backward`** (typical `loss.backward()` / `tensor.backward()`) — underapprox: any `Call` whose callee ends with `.backward` or bare name `backward` if bound in a way we can see cheaply; prefer `.backward` attribute calls.
2. A **`DataLoader(...)`** (or `torch.utils.data.DataLoader`) construction.
3. Optional later: configured entry-point list — **not required for v1** except where a rule below defines its own options.

**Skipped paths:** same family as python plugin defaults where applicable — `.pyi`, `test_*.py` / `*_test.py`, path segments `tests` / `__tests__` / `fixtures`, `conftest.py`, `__pycache__` — unless a rule explicitly needs tests (determinism-test-required reads tests on purpose).

Do **not** attempt whole-program “reachable from CLI” analysis. Local file evidence only.

### 1. `ml/require-global-seed` — recommended **`error`**

**Intent:** A training module must seed framework RNGs **before** the first model or DataLoader construction that the seed is meant to cover.

**Must seed when the module trains:**

- `torch.manual_seed` (or `torch.cuda.manual_seed_all` alone is **not** enough — need `torch.manual_seed` or equivalent `manual_seed` on torch module bind)
- If the file imports `random` (stdlib) in a way that is used or simply imported — require `random.seed(...)`. Underapprox: if `random` module is imported, require `random.seed`; if not imported, do not demand it.
- If the file imports `numpy` / `numpy.random` / `np` — require `numpy.random.seed` or `np.random.seed` (or `default_rng` construction with explicit seed **only if** we can see it easily; otherwise stick to `.seed`).

**Ordering:** flag if any required seed call appears **textually/AST-order after** the first `DataLoader(...)` construction or first clear model construction (`nn.Module` subclass instantiation is hard — use underapprox: first `DataLoader(` or first `torch.nn` module construction patterns we can prove, or first `.parameters(`/`Module()` if cheap). Minimum bar: seeds must appear before first `DataLoader(...)` when a DataLoader exists; if no DataLoader, before first `.backward()` is too late for weight init — prefer “before first clearly seeded-consumer construction”. Practical lock:

- Collect earliest AST position among: `DataLoader(...)` calls, and calls that construct modules when easy (`torch.nn.*` call callees).
- Collect positions of satisfactory seed calls.
- If training module and no satisfactory seed set → report on module top / first training evidence node.
- If seeds exist but all seed calls are after the earliest consumer → report ordering on the late seed or the early consumer.

**Quiet:** non-training modules; cannot prove training; seeds present in correct order for proven consumers.

**Suggestion:** concrete — call `torch.manual_seed` (and `random.seed` / `np.random.seed` as needed) before loaders/model init.

### 2. `ml/seed-must-reach-framework-rng` — recommended **`error`**

**Intent:** Catch the kevin failure mode: a `seed` parameter or CLI dest exists and is consumed only by split / `train_test_split` / local `random.Random(seed)`, while the module also trains — seed never reaches framework RNG.

**Detection (underapprox):**

- In a **training module**, find parameters / argparse destinations / variables named `seed` or `*_seed` or `split_seed` (configurable later; v1 hardcode common names: `seed`, `train_seed`, `split_seed`).
- If such a value is **used** and **none** of the uses flow (same-function simple) into `torch.manual_seed` / `torch.cuda.manual_seed` / `np.random.seed` / `random.seed` / `Generator(seed=...)` / `DataLoader(..., generator=...)`, report.
- If walker cannot follow the value through a call → **silence** (underapprox).
- If no seed-like name exists → silence (covered by require-global-seed instead).

**Suggestion:** pass the seed into `torch.manual_seed` / numpy / random / loader `generator`.

### 3. `ml/dataloader-worker-seeding` — recommended **`error`**

**Intent:** `DataLoader(num_workers=N)` with `N > 0` (static int/constant) and neither `worker_init_fn=` nor `generator=` set → report.

**Quiet:** `num_workers` missing / 0 / non-static; both or either of `worker_init_fn` / `generator` present; non-DataLoader calls.

**Suggestion:** add `worker_init_fn` and/or `generator=` seeded from the run seed.

### 4. `ml/tf32-must-be-explicit` — recommended **`error`**

**Intent:** A module that moves a model/tensor to CUDA must set both:

- `torch.backends.cuda.matmul.allow_tf32 = …`
- `torch.backends.cudnn.allow_tf32 = …`

(or equivalent attribute assigns we can see) **somewhere in that module** (same file), rather than inheriting Ampere defaults silently.

**CUDA move evidence (underapprox):** `.cuda()` call, or `.to("cuda"|"cuda:N"|device("cuda"))`, or `device="cuda"` literal in call kwargs. Do not require proving the object is an `nn.Module`.

**Quiet:** no CUDA move in file; both assigns present (any bool value — explicit false or true both OK); cannot prove CUDA move.

**Suggestion:** set both `allow_tf32` flags explicitly next to device setup.

### 5. `ml/determinism-test-required` — recommended **`error`**

**Intent:** A training **entry point** must have a test that trains twice and asserts identical weights (static presence, not execution).

**Entry point (v1 lock):**

- Rule option `entryPoints: string[]` optional list of dotted or path-style names.
- Plus heuristic: module-level functions named `train` / `main` in a **training module**, or `if __name__ == "__main__"` block that calls such a function in a training module.
- If neither options nor heuristic hits → no report.

**Test presence (underapprox, cousin of `python/public-exports-tested`):**

- Some test file references the entry point symbol (import / attribute) **and** the test module text/AST shows a pattern of **two** train invocations and a weight/tensor equality assert (`assertEqual`, `torch.equal`, `allclose`, `==` on state_dict keys, etc.). Perfect semantic proof not required — require:
  - reference to entry point, and
  - at least two calls to a train-like name or the entry point, and
  - an assert involving equality/allclose/equal on tensors or `state_dict`.
- If entry point found and no such test → report on entry point def / main guard.

**Options schema:** `{ entryPoints?: string[] }` via normal rule options. Empty/default = heuristic only.

**Suggestion:** add a test that runs the entry point twice under a fixed seed and asserts identical weights.

### 6. `ml/deterministic-algorithms-opt-in` — recommended **`off`**

**Intent:** Opt-in only. When enabled, a training module must call `torch.use_deterministic_algorithms(...)` or set `torch.backends.cudnn.deterministic = True` (attribute assign).

**Quiet when recommended off.** When user enables error/warn: training module without either → report.

**Suggestion:** call `torch.use_deterministic_algorithms(True)` or set `cudnn.deterministic` explicitly (document speed tradeoff in docs, not in every message spam).

## Plugin export surface

```ts
// packages/ml/src/index.ts
rules: {
  "require-global-seed": ...,
  "seed-must-reach-framework-rng": ...,
  "dataloader-worker-seeding": ...,
  "tf32-must-be-explicit": ...,
  "determinism-test-required": ...,
  "deterministic-algorithms-opt-in": ...,
},
configs: {
  recommended: {
    rules: {
      "ml/require-global-seed": "error",
      "ml/seed-must-reach-framework-rng": "error",
      "ml/dataloader-worker-seeding": "error",
      "ml/tf32-must-be-explicit": "error",
      "ml/determinism-test-required": "error",
      "ml/deterministic-algorithms-opt-in": "off",
    },
  },
},
```

No `provides`. No `ruff` / `biome` section required.

## Tests / fixtures

- Per-rule: at least one **bad** fixture (expect violations with correct `ruleId`) and one **ok** fixture (exit clean for that rule).
- Fixture configs list `@qualety/python` + `@qualety/ml` as needed.
- Plugin index test: name, rule keys, recommended map including opt-in **off**.
- Do not enable ml on root monorepo qualety.config.

## Docs

- `docs/rulesets/ml.md` — full honest catalog for the six rules.
- Companion closure pair.
- Touch SPECS plugin list / “what we own” only if there is an existing bullet list of official plugins — add ml one-liner + the six ids. Do not expand scope into #118+.

## Acceptance

- [ ] `@qualety/ml` package builds and is loadable via official loader
- [ ] Six rules implemented with locked recommended severities
- [ ] Fixtures + unit tests green in CI
- [ ] `docs/rulesets/ml.md` + companion pair; docs-export honesty satisfied
- [ ] No root dogfood ml; no init auto-add; no release
- [ ] Underapprox silence paths covered by ok fixtures / unit cases where practical
- [ ] PR body detailed squash quality; `Closes #117`

## Related

- Siblings (not this PR): #118, #119, #120, #121, #123
- Python provider: `@qualety/python`
- Motivating project: NavehBrenner/kevin

**Comment by NavehBrenner** [comment]
/oc plan

Treat the **issue body** of #117 as the sole source of truth (locked 2026-09-02). Plan implementation of `@qualety/ml` package scaffold + the six determinism/seeding rules exactly as specified there. Do not expand into #118–#121 or #123.

**Comment by NavehBrenner** [comment]
/oc build the plan above

Treat the plan comment https://github.com/NavehBrenner/qualety/issues/117#issuecomment-5507179753 and the **issue body** of #117 as source of truth.
